Backport remove duplicate delegators fix to the 3.21.x branch

settermjd opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This PR backports #251 to the 3.21.x branch. The reason for doing so is because this is the latest release that is supported by a number of other laminas projects, such as laminas-inputfilter. Given that, it makes sense to me to backport the fix here, so that it's available until the other packages support the 4.4.x release, where the fix was originally applied.

settermjd committed:
Remove duplicate delgators for all services

The mergeDelegators' documentation says that it merges delegators,
avoiding referencing the same delegators for the same service. However,
from experimenting with it, duplicates were allowed.  So, I've created
this small patch to ensure that duplicates are removed, along with a
covering test for it to verify the change.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Remove unused call to ReflectionProperty::setAccessible

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
I can merge the PR if that's okay.

arhimede commented:
LGTM

arhimede commented:
> I can merge the PR if that's okay.

go ahead

gsteel commented:
FYI, there's no need for manual PRs on the 3.22 and 3.23 branches, once merged and released, auto releases will create a merge-up pr for 3.22.x, that can be merged and a patch release of 3.22 can go out, which will create another merge-up pr for 3.23.

If auto releases fails, just cut the merge-up branch yourself from the tip of the released branch and raise that as a PR against the next feature branch…

Doing it this way keeps the older branches merged into the newer branches preventing future conflicts 👍

gsteel reviewed (changes requested):
Can you please do a force-push here. CI did not run for some reason.

settermjd commented:
> Can you please do a force-push here. CI did not run for some reason.

Shall do, when I get home.

settermjd commented:
> FYI, there's no need for manual PRs on the 3.22 and 3.23 branches, once merged and released, auto releases will create a merge-up pr for 3.22.x, that can be merged and a patch release of 3.22 can go out, which will create another merge-up pr for 3.23.
> 
> If auto releases fails, just cut the merge-up branch yourself from the tip of the released branch and raise that as a PR against the next feature branch…
> 
> Doing it this way keeps the older branches merged into the newer branches preventing future conflicts 👍

Awesome. Thanks for the info. Will sort this out.

settermjd commented:
> Can you please do a force-push here. CI did not run for some reason.

I tried, but got nowhere.

settermjd committed:
Add missing import for ReflectionProperty

I'm not sure why this was missed, but am adding it now so that there are
no issues.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Properly sort the use statements

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

samsonasik commented on test/ServiceManagerTest.php:
```diff
@@ -6,6 +6,7 @@
 
 use DateTime;
 use Laminas\ServiceManager\ConfigInterface;
+use Laminas\ContainerConfigTest\TestAsset\DelegatorFactory;
```
```
use Laminas\ContainerConfigTest\TestAsset\DelegatorFactory;
```

should be before

```
use Laminas\ServiceManager\ConfigInterface;
```

run:

```
composer cs-fix 
```

should handle it.

settermjd replied:
I thought that I pushed just such a change earlier. Thanks for the spot though.

settermjd commented:
This PR depends on #262.

settermjd commented:
Just force pushed the four commits after amending them to sign them all off.

settermjd commented:
#262 has been merged into this PR. Can it now be merged?

settermjd commented:
Thanks kindly, @gsteel.
